rest/oidc_api_test.go

```diff
@@ -1,139 +1,816 @@
+			require.NoError(t, err, "Error sending request")
+			if assertWantError(t, response, tc.wantError, grantTypeRefreshToken) {
+				return
+			}
+			require.Equal(t, http.StatusOK, response.StatusCode)
+
+			// Validate received token refresh response.
+			require.NoError(t, err, json.NewDecoder(response.Body).Decode(&receivedToken))
+			require.NoError(t, response.Body.Close(), "Error closing response body")
+			wantTokenResponse = mockAuthServer.options.tokenResponse
+			assert.NotEmpty(t, receivedToken.SessionID, "session_id doesn't exist")
+			assert.Equal(t, "foo_noah", receivedToken.Username, "name mismatch")
+			assert.Equal(t, wantTokenResponse.IDToken, receivedToken.IDToken, "id_token mismatch")
+			assert.Equal(t, wantTokenResponse.RefreshToken, receivedToken.RefreshToken, "refresh_token mismatch")
+			if tc.providers["foo"].IncludeAccessToken {
+				assert.Equal(t, wantTokenResponse.AccessToken, receivedToken.AccessToken, "access_token mismatch")
+				assert.Equal(t, wantTokenResponse.TokenType, receivedToken.TokenType, "token_type mismatch")
+				assert.Equal(t, wantTokenResponse.Expires, receivedToken.Expires, "expires_in mismatch")
+			}
+			// Query db endpoint with Bearer token
+			request, err = http.NewRequest(http.MethodGet, dbEndpoint, nil)
+			require.NoError(t, err, "Error creating new request")
+			request.Header.Add("Authorization", receivedToken.IDToken)
+			response, err = http.DefaultClient.Do(request)
+			require.NoError(t, err, "Error sending request with bearer token")
+			require.Equal(t, http.StatusOK, response.StatusCode)
+			require.NoError(t, err, json.NewDecoder(response.Body).Decode(&responseBody))
+			require.NoError(t, response.Body.Close(), "Error closing response body")
+			assert.Equal(t, restTester.DatabaseConfig.Name, responseBody["db_name"])
+		})
+	}
+}
+
+// assertHttpResponse asserts the wantError against HTTP response.
+func assertHttpResponse(t *testing.T, response *http.Response, wantError wantError) {
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	require.NoError(t, err, "error reading response body")
+	assert.Contains(t, string(bodyBytes), wantError.message)
+	assert.Equal(t, wantError.code, response.StatusCode)
+	require.NoError(t, response.Body.Close(), "error closing response body")
+}
+
+// refreshProviderConfig updates the issuer URL and sets the discovery endpoint
+// in provider configuration.
+func refreshProviderConfig(providers auth.OIDCProviderMap, issuer string) {
+	for name, provider := range providers {
+		provider.Issuer = issuer + "/" + name
+		provider.DiscoveryURI = provider.Issuer + auth.OIDCDiscoveryConfigPath
+	}
+}
+
+// assertWantError asserts the expected error kind, status code and messages against the actual response.
+// Return true if the wantError matches with the error in response and false otherwise.
+func assertWantError(t *testing.T, response *http.Response, wantError wantError, grantType grantType) (exit bool) {
```

Comment:
We're having to jump through hoops here because the tests are trying to evaluate two different things (the error on an auth request, and the error on a refresh request) using the same test parameter (wantError).  

It would be cleaner to have the test define two parameters (expectedAuthError, expectedRefreshError), so that the test can just do assertHttpResponse(t, response, expected***Error) instead of this switch and grant code condition processing.

The main concern I have with this level of unnecessary complexity in the test code is that it increases the risk of bugs in the test code, which then can result in missed bugs in the actual code we're trying to test.

Reply:
Added forceAuthError, forceRefreshError fields in the test struct and removed the switch conditional processing.